sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/WebExceptionRetryPolicy.java

```diff
@@ -41,79 +37,74 @@ public WebExceptionRetryPolicy(RetryContext retryContext) {
+        if (this.isOutOfRetries()) {
+            logger
+                .warn(
+                    "WebExceptionRetryPolicy() No more retrying on endpoint {}, operationType = {}, count = {}, " +
+                        "isAddressRefresh = {}",
+                    this.locationEndpoint,
+                    this.request.getOperationType(),
+                    this.retryCount,
+                    this.request.isAddressRefresh());
+
             this.durationTimer.stop();
             return Mono.just(ShouldRetryResult.noRetry());
         }
 
-        // Received Connection error (HttpRequestException), initiate the endpoint rediscovery
-        CosmosException webException = Utils.as(e, CosmosException.class);
-        if (WebExceptionUtility.isNetworkFailure(e) && this.isReadRequest &&
-            (webException != null && WebExceptionUtility.isReadTimeoutException(webException) &&
-                Exceptions.isSubStatusCode(webException, HttpConstants.SubStatusCodes.GATEWAY_ENDPOINT_READ_TIMEOUT))) {
 
-            // if operationType AddressRefresh then just retry
-            if (this.request.isAddressRefresh()) {
-                return shouldRetryAddressRefresh();
+        if (WebExceptionUtility.isNetworkFailure(e)) {
+            if (this.isReadRequest || WebExceptionUtility.isWebExceptionRetriable(e)) {
+                int delayInSeconds = this.timeoutPolicy.getTimeoutAndDelaysList().get(this.retryCount).getDelayForNextRequestInSeconds();
+                this.request.setResponseTimeout(this.timeoutPolicy.getTimeoutAndDelaysList().get(this.retryCount).getResponseTimeout());
+                // Increase the retry count after calculating the delay
+                retryCount++;
+                if (logger.isDebugEnabled()) {
+                    logger
+                        .warn("WebExceptionRetryPolicy() Retrying on endpoint {}, operationType = {}, count = {}, " +
+                                "isAddressRefresh = {}, shouldForcedAddressRefresh = {}, " +
+                                "shouldForceCollectionRoutingMapRefresh = {}",
+                            this.locationEndpoint, this.request.getOperationType(), this.retryCount,
+                            this.request.isAddressRefresh(),
+                            this.request.shouldForceAddressRefresh(),
+                            this.request.forceCollectionRoutingMapRefresh);
+                }
+
+                return Mono.just(ShouldRetryResult.retryAfter(Duration.ofSeconds(delayInSeconds)));
             }
-
-            return Mono.just(ShouldRetryResult.retryAfter(Duration.ofSeconds(retryDelay)));
         }
 
-        if (!WebExceptionUtility.isWebExceptionRetriable(e)) {
-            // Have caller propagate original exception.
-            this.durationTimer.stop();
-            return Mono.just(ShouldRetryResult.noRetryOnNonRelatedException());
+        if (logger.isDebugEnabled()) {
+            logger
+                .warn(
```

Comment:
Same here.

Reply:
We should have warn there, and not the guarding if condition of debugEnabled. I have removed that.